hubble/cmd/common/conn/conn.go

```diff
@@ -30,13 +34,100 @@ var GRPCOptionFuncs []GRPCOptionFunc
 func init() {
 	GRPCOptionFuncs = append(
 		GRPCOptionFuncs,
-		grpcInterceptors,
+		grpcUnaryInterceptors,
+		grpcStreamInterceptors,
 		grpcOptionTLS,
 	)
 }
 
-func grpcInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
-	return grpc.WithUnaryInterceptor(timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout))), nil
+func grpcUnaryInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainUnaryInterceptor(
+		timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout)),
+		logVersionMismatchUnaryInterceptor(),
+	)
+	return option, nil
+}
+
+func grpcStreamInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainStreamInterceptor(
+		logVersionMismatchStreamInterceptor(),
+	)
+	return option, nil
+}
+
+func logVersionMismatchUnaryInterceptor() grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req any, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		var header metadata.MD
+		opts = append(opts, grpc.Header(&header))
+		if err := invoker(ctx, method, req, reply, cc, opts...); err != nil {
+			return err
+		}
+		logVersionMismatch(header)
+		return nil
+	}
+}
+
+func logVersionMismatchStreamInterceptor() grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		stream, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, err
+		}
+		// stream.Header() blocks until metadata is ready to read.
+		// This could be forever if no metadata is ever sent, after
+		// the first read from the stream or after an explicit call
+		// to SendHeader() server-side.
+		go func() {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			header, err := stream.Header()
+			if err == nil {
+				logVersionMismatch(header)
+			}
+		}()
+		return stream, nil
+	}
+}
+
+func logVersionMismatch(header metadata.MD) {
+	var relayVersion string
+	relayVersionHeaders := header.Get(relaydefaults.GRPCMetadataRelayVersionKey)
+	if len(relayVersionHeaders) > 0 {
+		relayVersion = relayVersionHeaders[0]
+	}
+	var serverVersion string
+	serverVersionHeaders := header.Get(serverdefaults.GRPCMetadataServerVersionKey)
+	if len(serverVersionHeaders) > 0 {
+		serverVersion = serverVersionHeaders[0]
+	}
+
+	relayVersionMissing := relayVersion == ""
+	relayVersionMismatch := relayVersion != build.RelayVersion.SemVer()
+	serverVersionMissing := serverVersion == ""
+	serverVersionMismatch := serverVersion != build.ServerVersion.SemVer()
+
+	if relayVersionMissing && serverVersionMissing {
+		logger.Logger.With(
```

Comment:
I tend to agree with the linter, that you shouldn't call the global logger, which might not even be initialized.  But given that this seems to be the way the CLI handles the logger and the global gRPC option, I don't want to suggest a larger refactor and I'd be OK with just silencing the linter for these.
But it would be nice if we could clean this up in the future.

Reply:
I agree we should clean this up.

Reply:
I think the linter warning was more because I was using With(), whereas with slog, we should pass the args directly to Debug/Warn/etc, or use With to obtain a new logger with the fields set.

In any case, I still agree that passing logger instances is probably better. For now, I've lifted the dependency up at the highest point when the interceptor is created so it's not hidden deeply in the implementation.